File: targets/DailyVocabWidget/DailyWordStore.swift
import Foundation

struct DailySnapshot: Codable, Equatable {
  let level: String
  let localDate: String
  let wordId: String
  let word: String
  let oneLiner: String
}

struct CatalogWord: Codable {
  let id: String
  let word: String
  let oneLiner: String
}

struct CatalogFile: Codable {
  let version: Int
  let level: String
  let words: [CatalogWord]
}

/// Shared with the React Native host via App Group `group.com.dailyvocab.app`.
enum DailyWordStore {
  static let suiteName = "group.com.dailyvocab.app"
  static let snapshotKey = "dailySnapshot"
  static let levelKey = "activeLevel"
  static let shownKey = "shownYearByWordId"

  static var defaults: UserDefaults {
    UserDefaults(suiteName: suiteName) ?? .standard
  }

  static func localDateString(now: Date = Date(), timeZone: TimeZone = .current) -> String {
    let formatter = DateFormatter()
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = timeZone
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter.string(from: now)
  }

  static func yearDigit(now: Date = Date()) -> Int {
    Calendar.current.component(.year, from: now) % 10
  }

  static func loadShownYears() -> [String: Int] {
    guard let json = defaults.string(forKey: shownKey),
          let data = json.data(using: .utf8),
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return [:]
    }
    var map: [String: Int] = [:]
    for (key, value) in object {
      if let number = value as? Int {
        map[key] = number
      } else if let number = value as? NSNumber {
        map[key] = number.intValue
      }
    }
    return map
  }

  static func saveShownYears(_ map: [String: Int]) {
    guard let data = try? JSONSerialization.data(withJSONObject: map),
          let json = String(data: data, encoding: .utf8) else { return }
    defaults.set(json, forKey: shownKey)
    defaults.synchronize()
  }

  static func loadSnapshot() -> DailySnapshot? {
    guard let json = defaults.string(forKey: snapshotKey),
          let data = json.data(using: .utf8) else { return nil }
    return try? JSONDecoder().decode(DailySnapshot.self, from: data)
  }

  static func saveSnapshot(_ snapshot: DailySnapshot) {
    guard let data = try? JSONEncoder().encode(snapshot),
          let json = String(data: data, encoding: .utf8) else { return }
    defaults.set(json, forKey: snapshotKey)
    defaults.synchronize()
  }

  static func loadCatalog(level: String, bundle: Bundle = .main) -> [CatalogWord] {
    if let url = bundle.url(forResource: level, withExtension: "json"),
       let data = try? Data(contentsOf: url),
       let file = try? JSONDecoder().decode(CatalogFile.self, from: data) {
      return file.words
    }
    return []
  }

  static func ensureTodaysWord(
    level: String,
    words: [CatalogWord],
    state: DailySnapshot?,
    now: Date = Date(),
    randomInt: (Int) -> Int = { Int.random(in: 0..<$0) }
  ) -> DailySnapshot? {
    let today = localDateString(now: now)
    if let state, state.localDate == today {
      var shown = loadShownYears()
      shown[state.wordId] = yearDigit(now: now)
      saveShownYears(shown)
      return state
    }
    guard !words.isEmpty else { return nil }

    var shown = loadShownYears()
    let current = yearDigit(now: now)
    let previous = (current + 9) % 10
    var pool = words.filter { word in
      guard let stamp = shown[word.id] else { return true }
      return stamp != current && stamp != previous
    }
    if pool.isEmpty {
      pool = words.filter { $0.id != state?.wordId }
      if pool.isEmpty { pool = words }
    }

    var index = randomInt(pool.count)
    if pool.count > 1, let previousId = state?.wordId {
      var guardCount = 0
      while pool[index].id == previousId && guardCount < 8 {
        index = randomInt(pool.count)
        guardCount += 1
      }
      if pool[index].id == previousId {
        index = (index + 1) % pool.count
      }
    }

    let entry = pool[index]
    shown[entry.id] = current
    saveShownYears(shown)
    return DailySnapshot(
      level: level,
      localDate: today,
      wordId: entry.id,
      word: entry.word,
      oneLiner: entry.oneLiner
    )
  }

  static func resolveForWidget(now: Date = Date(), bundle: Bundle = .main) -> DailySnapshot? {
    let existing = loadSnapshot()
    let today = localDateString(now: now)
    if let existing, existing.localDate == today {
      var shown = loadShownYears()
      shown[existing.wordId] = yearDigit(now: now)
      saveShownYears(shown)
      return existing
    }
    let level = defaults.string(forKey: levelKey) ?? existing?.level ?? "beginner"
    let words = loadCatalog(level: level, bundle: bundle)
    guard let next = ensureTodaysWord(level: level, words: words, state: existing, now: now) else {
      return existing
    }
    if next != existing {
      saveSnapshot(next)
    }
    return next
  }

  static func nextMidnight(from date: Date = Date(), calendar: Calendar = .current) -> Date {
    let start = calendar.startOfDay(for: date)
    return calendar.date(byAdding: .day, value: 1, to: start) ?? date.addingTimeInterval(86_400)
  }
}

File: modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/DailyWordStore.kt
package expo.modules.widgetbridge

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.random.Random

data class DailySnapshot(
  val level: String,
  val localDate: String,
  val wordId: String,
  val word: String,
  val oneLiner: String,
)

data class CatalogWord(
  val id: String,
  val word: String,
  val oneLiner: String,
)

object DailyWordStore {
  const val PREFS = "dayink_widget"
  const val SNAPSHOT_KEY = "dailySnapshot"
  const val LEVEL_KEY = "activeLevel"
  const val SHOWN_KEY = "shownYearByWordId"

  fun localDateString(now: Date = Date(), timeZone: TimeZone = TimeZone.getDefault()): String {
    val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US)
    fmt.timeZone = timeZone
    return fmt.format(now)
  }

  fun yearDigit(now: Date = Date()): Int {
    val cal = Calendar.getInstance()
    cal.time = now
    return cal.get(Calendar.YEAR) % 10
  }

  fun loadShownYears(context: Context): MutableMap<String, Int> {
    val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
      .getString(SHOWN_KEY, null) ?: return mutableMapOf()
    return runCatching {
      val obj = JSONObject(raw)
      val map = mutableMapOf<String, Int>()
      val keys = obj.keys()
      while (keys.hasNext()) {
        val key = keys.next()
        map[key] = obj.getInt(key)
      }
      map
    }.getOrDefault(mutableMapOf())
  }

  fun saveShownYears(context: Context, map: Map<String, Int>) {
    val obj = JSONObject()
    for ((key, value) in map) {
      obj.put(key, value)
    }
    context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
      .edit()
      .putString(SHOWN_KEY, obj.toString())
      .commit()
  }

  fun loadSnapshot(context: Context): DailySnapshot? {
    val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
      .getString(SNAPSHOT_KEY, null) ?: return null
    return runCatching {
      val obj = JSONObject(raw)
      DailySnapshot(
        level = obj.getString("level"),
        localDate = obj.getString("localDate"),
        wordId = obj.getString("wordId"),
        word = obj.getString("word"),
        oneLiner = obj.getString("oneLiner"),
      )
    }.getOrNull()
  }

  fun saveSnapshot(context: Context, snapshot: DailySnapshot) {
    val obj = JSONObject()
      .put("level", snapshot.level)
      .put("localDate", snapshot.localDate)
      .put("wordId", snapshot.wordId)
      .put("word", snapshot.word)
      .put("oneLiner", snapshot.oneLiner)
    context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
      .edit()
      .putString(SNAPSHOT_KEY, obj.toString())
      .commit()
  }

  fun loadCatalog(context: Context, level: String): List<CatalogWord> {
    val assetName = "words/$level.json"
    return runCatching {
      context.assets.open(assetName).bufferedReader().use { it.readText() }
        .let { JSONObject(it) }
        .getJSONArray("words")
        .toWordList()
    }.getOrDefault(emptyList())
  }

  fun ensureTodaysWord(
    context: Context,
    level: String,
    words: List<CatalogWord>,
    state: DailySnapshot?,
    now: Date = Date(),
    randomInt: (Int) -> Int = { Random.nextInt(it) },
  ): DailySnapshot? {
    val today = localDateString(now)
    val digit = yearDigit(now)
    if (state != null && state.localDate == today) {
      val shown = loadShownYears(context)
      shown[state.wordId] = digit
      saveShownYears(context, shown)
      return state
    }
    if (words.isEmpty()) return null

    val shown = loadShownYears(context)
    val previousDigit = (digit + 9) % 10
    var pool = words.filter { word ->
      val stamp = shown[word.id]
      stamp == null || (stamp != digit && stamp != previousDigit)
    }
    if (pool.isEmpty()) {
      pool = words.filter { it.id != state?.wordId }
      if (pool.isEmpty()) pool = words
    }

    var index = randomInt(pool.size)
    val previous = state?.wordId
    if (pool.size > 1 && previous != null) {
      var guard = 0
      while (pool[index].id == previous && guard < 8) {
        index = randomInt(pool.size)
        guard += 1
      }
      if (pool[index].id == previous) {
        index = (index + 1) % pool.size
      }
    }

    val entry = pool[index]
    shown[entry.id] = digit
    saveShownYears(context, shown)
    return DailySnapshot(
      level = level,
      localDate = today,
      wordId = entry.id,
      word = entry.word,
      oneLiner = entry.oneLiner,
    )
  }

  fun resolveForWidget(context: Context): DailySnapshot? {
    val existing = loadSnapshot(context)
    val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    val today = localDateString()
    if (existing != null && existing.localDate == today) {
      val shown = loadShownYears(context)
      shown[existing.wordId] = yearDigit()
      saveShownYears(context, shown)
      return existing
    }
    val level = prefs.getString(LEVEL_KEY, null) ?: existing?.level ?: return null
    val words = loadCatalog(context, level)
    val next = ensureTodaysWord(context, level, words, existing) ?: return existing
    if (next.wordId != existing?.wordId || next.localDate != existing.localDate) {
      saveSnapshot(context, next)
    }
    return next
  }
}

private fun JSONArray.toWordList(): List<CatalogWord> {
  val list = ArrayList<CatalogWord>(length())
  for (i in 0 until length()) {
    val obj = getJSONObject(i)
    list.add(
      CatalogWord(
        id = obj.getString("id"),
        word = obj.getString("word"),
        oneLiner = obj.getString("oneLiner"),
      ),
    )
  }
  return list
}
